### PR TITLE
Implement max character limits on application fields #162592036

### PR DESCRIPTION
**Pull request by chrisdolendo**
https://www.pivotaltracker.com/story/show/162592036

### CODE REVIEW

**Comment by chrisdolendo** [comment]
Code climate complaints don't really make sense so permission to mark as Fixed: https://codeclimate.com/github/Exygy/sf-dahlia-lap/pull/204

**Comment by akegan** [comment]
@chrisdolendo heads up - unit tests are failing

**Comment by akegan** [comment]
labeled all of the code climate flags as won't fix

File: app/javascript/components/applications/application_form/AlternateContactSection.js [hunk 1]
```diff
@@ -80,21 +84,26 @@ const AlternateContactSection = ({editValues}) => {
               </div>
               <div className='small-4 columns'>
                 <label>Alternate Contact Type Other</label>
-                <Text field='alternate_contact_type_other' />
+                <Text field='alternate_contact_type_other' maxLength={maxLengthMap['alternate_contact_type_other']} />
```
Comment (akegan) [inline app/javascript/components/applications/application_form/AlternateContactSection.js]: one minor note just on formatting, I'd prefer if we broke these one-line statements into multiple lines to be consistent with the way we format other components

File: spec/javascript/e2e/ApplicationNewPage.e2e.js [hunk 2]
```diff
@@ -27,8 +30,8 @@ describe('ApplicationNewPage', () => {
     await sharedSteps.goto(page, `/listings/${NON_LEASE_UP_LISTING_ID}/applications/new`)
 
     // Enter in required information
-    await page.type('#first_name', FIRST_NAME)
-    await page.type('#last_name', LAST_NAME)
+    await page.type('#first_name', LONG_FIRST_NAME)
+    await page.type('#last_name', LONG_LAST_NAME)
```
Comment (akegan) [inline spec/javascript/e2e/ApplicationNewPage.e2e.js]: what is being tested here? that the whole string can be typed in?

File: app/javascript/utils/formUtils.js [hunk 3]
```diff
@@ -40,6 +40,28 @@ const submitErrors = (formApi) => {
   return formApi.submits > 0 ? formApi.errors : mapValues(formApi.errors, () => null)
 }
 
+export const maxLengthMap = {
+  first_name: 40,
+  middle_name: 20,
+  last_name: 40,
+  day: 2,
+  month: 2,
+  year: 4,
```
Comment (akegan) [inline app/javascript/utils/formUtils.js]: are you using these date validations anywhere?

**Review by akegan — CHANGES_REQUESTED** [review]
one minor request and a couple of questions

File: app/javascript/utils/formUtils.js (hunk 3 above)
Comment (chrisdolendo) [reply app/javascript/utils/formUtils.js]: Ah yes, the dates already have validation, removing these.

File: spec/javascript/e2e/ApplicationNewPage.e2e.js (hunk 2 above)
Comment (chrisdolendo) [reply spec/javascript/e2e/ApplicationNewPage.e2e.js]: `LONG_FIRST_NAME` is a string over the 40 char limit for the input. The test makes sure that even if we type in `LONG_FIRST_NAME` that `FIRST_NAME` (which is the truncated string which at the max limit) is the value that actually gets saved.

Comment (chrisdolendo) [reply spec/javascript/e2e/ApplicationNewPage.e2e.js]: Added a comment to clarify :)

File: app/javascript/components/applications/application_form/AlternateContactSection.js (hunk 1 above)
Comment (chrisdolendo) [reply app/javascript/components/applications/application_form/AlternateContactSection.js]: Done!

File: spec/javascript/e2e/ApplicationNewPage.e2e.js (hunk 2 above)
Comment (akegan) [reply spec/javascript/e2e/ApplicationNewPage.e2e.js]: maybe it'd be clearer if FIRST_NAME and LAST_NAME were TRUNCATED_FIRST_NAME and TRUNCATED_LAST_NAME

**Comment by akegan** [comment]
just one more request on the e2e tests. I'll start QAing now

File: app/javascript/utils/formUtils.js [hunk 4]
```diff
@@ -44,9 +44,6 @@ export const maxLengthMap = {
   first_name: 40,
   middle_name: 20,
   last_name: 40,
-  day: 2,
```
Comment (akegan) [inline app/javascript/utils/formUtils.js]: I think we should actually keep these after all and use them. Sorry for the flip-flop!

**Comment by akegan** [comment]
QA looks good! Here's the application I created: https://dahlia-lap-full-pr-204.herokuapp.com/applications/a0o0t000000ltf7AAA?showAddBtn=true

File: app/javascript/utils/formUtils.js (hunk 4 above)
Comment (chrisdolendo) [reply app/javascript/utils/formUtils.js]: done!

File: spec/javascript/e2e/ApplicationNewPage.e2e.js (hunk 2 above)
Comment (chrisdolendo) [reply spec/javascript/e2e/ApplicationNewPage.e2e.js]: done

**Comment by akegan** [comment]
@chrisdolendo snapshot tests failing but other than that code looks good. QAing now

**Comment by akegan** [comment]
QA looks good too! let me know when you update snapshots so I can check out the changes then you're good to merge!

**Comment by chrisdolendo** [comment]
@akegan snapshots are updated!

**Review by akegan — APPROVED** [review]
looks great! so excited for this change!

**Pull request merged**